curl/tests/request_test.cc

```diff
@@ -0,0 +1,116 @@
+#include <catch.hpp>
+#include "mock_curl.hpp"
+#include <leatherman/curl/client.hpp>
+#include <leatherman/curl/request.hpp>
+
+using namespace std;
+
+namespace leatherman { namespace curl {
+
+    struct mock_client : client {
+        curl_handle const& get_handle() { return client::get_handle(); }
+    };
+
+    TEST_CASE("curl::request") {
+        mock_client test_client;
+
+        SECTION("Headers should be addable and retrievable from the request") {
+            request test_request {"http://valid.com"};
+            test_request.add_header("header_name", "header_value");
+            auto header = test_request.header("header_name");
+            REQUIRE(header);
+            REQUIRE(*(header) == "header_value");
+        }
+
+        SECTION("Headers should be removable from the request") {
+            request test_request {"http://valid.com"};
+            test_request.add_header("header_name", "header_value");
+            test_request.remove_header("header_name");
+            auto header = test_request.header("header_name");
+            REQUIRE(header == nullptr);
+        }
+
+        SECTION("Headers should be enumerable") {
+            int i = 0;
+            string expected_name, expected_value;
+            request test_request {"http://valid.com"};
+            test_request.add_header("header_0", "header_value_0");
+            test_request.add_header("header_1", "header_value_1");
+            test_request.add_header("header_2", "header_value_2");
+
+            test_request.each_header([&](string const& name, string const& value) {
+                expected_name  = "header_" + to_string(i);
+                expected_value = "header_value_" + to_string(i);
+                REQUIRE(name == expected_name);
+                REQUIRE(value == expected_value);
+                ++i;
+                return true;
+            });
+        }
+
+        SECTION("Cookies should be present in the request when added") {
+            request test_request {"http://valid.com"};
```

Comment:
If you split the sections up, with those using the client in a new TEST_CASE, you could pull a lot of setup for these tests out before the sections as well.